### PR TITLE
Streamline memory management interface

### PR DESCRIPTION
**Pull request by j-hui**
@sedwards-lab is this something closer to what you had in mind?

I haven't updated any of the tests yet but the library builds successfully, so at least these changes are self-consistent.

### CODE REVIEW

**Review by sedwards-lab — APPROVED** [review]
I like seeing lots of code go away. LGTM

File: include/ssm.h
Comment (j-hui) [inline include/ssm.h:95]: Changed: All values are 32-bits now, regardless of architecture.

File: include/ssm.h
Comment (j-hui) [inline include/ssm.h]: Wart: we need to explicitly pad this because otherwise the compiler would do this for us inside of a `struct ssm_mm` (only on 64-bit archs), throwing off size calculations.

File: runtests.sh
Comment (j-hui) [inline runtests.sh:70]: Added: Add list tests to the test script (I forgot to do so before)

File: include/ssm.h
Comment (j-hui) [inline include/ssm.h]: Wart: I made this a separate struct to make it slightly easier to compute the size, but the padding issue makes me think twice about this choice.

File: examples/list.c
Comment (j-hui) [inline examples/list.c:193]: Changed: now representing payload-free variants with just an `ssm_word_t`.

File: examples/list.c
Comment (j-hui) [inline examples/list.c]: Wart: Because the "heapness" of a value is no longer type-dependent, we need to check each value we want to drop or dup. Though perhaps not entirely a wart, since here we're trading off a little bit of runtime efficiency for massive memory efficiency.

File: examples/list.c
Comment (j-hui) [inline examples/list.c:196]: Added: the size of the initial list in the list example is now parameterized by the command line argument.

**Review by j-hui — COMMENTED** [review]
This is a fairly large diff, so I highlighted the additions and changes, in addition to some warts that might be worth discussing before landing this.

File: examples/list.c
Comment (j-hui) [reply examples/list.c]: I "fixed" this by hiding the check inside `ssm_{dup,drop,reuse}`, which are now macros. Under the hood, they call `ssm_{dup,drop,reuse}_unsafe`, which can be directly called instead when the caller is certain that the value points to the heap.

Comment (j-hui) [reply examples/list.c]: This should also make it easier to generate (slightly less efficient) code, and then optimize later by inlining with knowledge of type information.

**Comment by j-hui** [comment]
I reverted the memory representation to the old scheme, where `struct ssm_mm` was embedded inside heap payloads (rather than the other way around), because it seems to more resilient and efficient in the face of struct padding.

That being said, I kept the other ergonomic improvements introduced in this PR, which make examples easier to read and thus code easier to generate.

**Comment by j-hui** [comment]
Merging this now so to unblock myself; we can always revisit the struct padding issue later.

**Pull request merged**
